### PR TITLE
Allow running under root on Linux when unshare is available

### PR DESCRIPTION
**Pull request by jameshilliard**
I've reworked the `unshare` parts from #23 so that we test once on initialization that `unshare` is useable when running as root on Linux. This should avoid any cryptic errors when `unshare` is unavailable.

I've also reworked the runner to execute the `postgres` binary directly as opposed to being backgrounded by `pg_ctl`. This has the advantage of preventing orphan `postgres` in the event that cleanup fails as the `postgres` process should effectively be a child of the java module which would get killed at the same time as the java module.

### CODE REVIEW

File: src/main/java/io/zonky/test/db/postgres/embedded/PreparedDbProvider.java [hunk 1]
```diff
@@ -123,7 +123,7 @@ public ConnectionInfo createNewDatabase() throws SQLException
     public DataSource createDataSourceFromConnectionInfo(final ConnectionInfo connectionInfo) throws SQLException
     {
         final PGSimpleDataSource ds = new PGSimpleDataSource();
-        ds.setPortNumber(connectionInfo.getPort());
+        ds.setPortNumbers(new int[] {connectionInfo.getPort()});
```
Comment (tomix26) [inline src/main/java/io/zonky/test/db/postgres/embedded/PreparedDbProvider.java:126]: This change is not possible because the library should be compatible with a wider range of driver versions. If I use it with an older version of the driver, it causes the following error: `java.lang.NoSuchMethodError: org.postgresql.ds.PGSimpleDataSource.setServerNames([Ljava/lang/String;)V`

File: src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java [hunk 2]
```diff
@@ -133,6 +129,28 @@
...
+                ));
+                final ProcessBuilder builder = new ProcessBuilder(command);
+                final Process process = builder.start();
+                BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                try {
+                    process.waitFor();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                if (process.exitValue() == 0 && br.readLine() != "0") {
```
Comment (tomix26) [inline src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java:148]: Unfortunately, when I tried it on Docker, it didn't work as expected. The `useUnshare` variable was always set to `false`. I also tried to run `unshare -U id -u` command manually in a docker container but I got `unshare failed: Operation not permitted` error. When I investigated the problem further, I came to the conclusion that Docker blocks the `unshare` command by default because it has caused a lot of local privilege escalations over time. So the only workaround I've found is to use `--privileged` option, which works but it doesn't seem to be an ideal solution.

Could you please investigate it further and find a suitable solution? Because without working on Docker is this change useless.

File: src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java [hunk 3]
```diff
@@ -133,6 +129,28 @@
         this.pgStartupWait = pgStartupWait;
         Objects.requireNonNull(this.pgStartupWait, "Wait time cannot be null");
 
+        if (SystemUtils.IS_OS_LINUX) {
+            int uid = (int) new com.sun.security.auth.module.UnixSystem().getUid();
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java:133]: From what I can tell it is, I've used this same code in other projects which target multiple java versions.

File: src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java (hunk 2 above)
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java:148]: Actually, I hadn't really targeted docker as the main use case for this change so much as throwaway testing VM's where root is the default user or for cases where one wants to test a java application that itself needs to run as root(for example some system utility than needs full privileges) on the host. This check is to ensure that `unshare` is only used when it is confirmed to function correctly so that confusing error messages don't get generated.

Basically the `unshare` command can be thought of as a lightweight docker alternative for changing the uid. I think `--privileged` is needed because this is somewhat like a [docker in docker](https://www.docker.com/blog/docker-can-now-run-within-docker/).

File: src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java [hunk 4]
```diff
@@ -104,6 +99,7 @@
     private volatile FileOutputStream lockStream;
     private volatile FileLock lock;
     private final boolean cleanDataDirectory;
+    private boolean useUnshare = false;
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java]: changed to static

File: src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java [hunk 5]
```diff
@@ -133,6 +129,28 @@
         this.pgStartupWait = pgStartupWait;
         Objects.requireNonNull(this.pgStartupWait, "Wait time cannot be null");
 
+        if (SystemUtils.IS_OS_LINUX) {
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java]: moved

File: src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java [hunk 6]
```diff
@@ -186,8 +204,8 @@ public DataSource getDatabase(String userName, String dbName) {
     public DataSource getDatabase(String userName, String dbName, Map<String, String> properties)
     {
         final PGSimpleDataSource ds = new PGSimpleDataSource();
-        ds.setServerName("localhost");
-        ds.setPortNumber(port);
+        ds.setServerNames(new String[] {"localhost"});
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java]: reverted, I had changed this just due to the deprecation warning that popped up

File: src/main/java/io/zonky/test/db/postgres/embedded/PreparedDbProvider.java (hunk 1 above)
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/embedded/PreparedDbProvider.java]: reverted

File: src/main/java/io/zonky/test/db/postgres/util/LinuxUtils.java [hunk 7]
```diff
@@ -85,4 +94,47 @@ private static String resolveDistributionName() {
             return null;
         }
     }
+
+    private static boolean unshareUseable() {
+        if (SystemUtils.IS_OS_LINUX) {
+            int uid;
+            try {
+                Class<?> c = Class.forName("com.sun.security.auth.module.UnixSystem");
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply src/main/java/io/zonky/test/db/postgres/util/LinuxUtils.java]: did a quick regression test, everything still looks good on my end
